imap/autocreate.c

```diff
@@ -93,330 +93,399 @@ static const char *get_script_name(const char *filename)
         return p + 1;
 }
 
+enum SieveFileType {
+    SIEVE_TMP_1 = 1,
+    SIEVE_TMP_2,
+    SIEVE_BC_TMP,
+    SIEVE_SCRIPT,
+    SIEVE_BC_SCRIPT,
+    SIEVE_DEFAULT,
+    SIEVE_BC_LINK,
+    SIEVE_NUM_FILE_TYPES
+};
+static struct sieve_scripts_info {
+    enum SieveFileType stype;
+    char *ext;
+} sieve_names[] = {
+    {SIEVE_TMP_1, ".script.NEW"},
+    {SIEVE_TMP_2, ".NEW"},
+    {SIEVE_BC_TMP, ".bc.NEW"},
+    {SIEVE_SCRIPT, ".script"},
+    {SIEVE_BC_SCRIPT, ".bc"},
+    {SIEVE_DEFAULT, ""},
+    {SIEVE_BC_LINK, ".bc"}
+};
+
+struct sieve_scripts {
+    char tmpname1[MAX_FILENAME];
+    char tmpname2[MAX_FILENAME];
+    char bctmpname[MAX_FILENAME];
+    char scriptname[MAX_FILENAME];
+    char bcscriptname[MAX_FILENAME];
+    char defaultname[MAX_FILENAME];
+    char bclinkname[MAX_FILENAME];
+};
+
+/*
+ * Given the path to the sieve script directory `script_dir` and the sievename,
+ * generate filenames and return it in the `fnames` structure.
+ *
+ * On Success, returns 0, else returns 1
+ */
+static int setup_sieve_filenames(const char *script_dir, const char *sievename,
+                                 struct sieve_scripts *fnames)
+{
+    int r = 0, i;
+
+    for (i = 0; i < SIEVE_NUM_FILE_TYPES; i++) {
+        struct sieve_scripts_info *info = &sieve_names[i];
+        int ret;
+        switch(info->stype) {
+            case SIEVE_TMP_1:
+                ret = snprintf(fnames->tmpname1, MAX_FILENAME, "%s/%s%s",
+                               script_dir, sievename, info->ext);
+                break;
+            case SIEVE_TMP_2:
+                ret = snprintf(fnames->tmpname2, MAX_FILENAME, "%s/%s%s",
+                               script_dir, sievename, info->ext);
+                break;
+            case SIEVE_BC_TMP:
+                ret = snprintf(fnames->bctmpname, MAX_FILENAME, "%s/%s%s",
+                               script_dir, sievename, info->ext);
+                break;
+            case SIEVE_SCRIPT:
+                ret = snprintf(fnames->scriptname, MAX_FILENAME, "%s/%s%s",
+                               script_dir, sievename, info->ext);
+                break;
+            case SIEVE_BC_SCRIPT:
+                ret = snprintf(fnames->bcscriptname, MAX_FILENAME, "%s/%s%s",
+                               script_dir, sievename, info->ext);
+                break;
+            case SIEVE_DEFAULT:
+                ret = snprintf(fnames->defaultname, MAX_FILENAME, "%s/%s%s",
+                               script_dir, sievename, info->ext);
+                break;
+            case SIEVE_BC_LINK:
+                /* For some reason bclink name has no directory. */
```

Comment:
This is because a relative symlink target is relative to the location of the symlink, and since the defaultname symlink is being created in the appropriate directory, its target can't also specify the directory (otherwise you'd get like "user/f/foo/default" pointing to "[user/f/foo/]user/f/foo/somescript.bc" and things would fall apart :) )